File: src/lib/ipc/events.ts
import { listen } from '@tauri-apps/api/event';
import type { Match, Phase, Round, Tournament, TeamStanding } from './types';

export const onTournamentUpdated = (cb: (t: Tournament) => void) =>
  listen<Tournament>('tournament-updated', (e) => cb(e.payload));

export const onTournamentLoaded = (cb: (t: Tournament) => void) =>
  listen<Tournament>('tournament-loaded', (e) => cb(e.payload));

export const onTournamentClosed = (cb: () => void) =>
  listen('tournament-closed', () => cb());

export const onMatchUpdated = (cb: (m: Match) => void) =>
  listen<Match>('match-updated', (e) => cb(e.payload));

export const onRoundGenerated = (cb: (r: Round) => void) =>
  listen<Round>('round-generated', (e) => cb(e.payload));

export const onPhaseChanged = (cb: (from: Phase, to: Phase) => void) =>
  listen<{ from: Phase; to: Phase }>('phase-changed', (e) =>
    cb(e.payload.from, e.payload.to)
  );

export const onStandingsUpdated = (cb: (s: TeamStanding[]) => void) =>
  listen<TeamStanding[]>('standings-updated', (e) => cb(e.payload));

export const onDirtyChanged = (cb: (dirty: boolean, path: string | null) => void) =>
  listen<{ dirty: boolean; path: string | null }>('dirty-changed', (e) =>
    cb(e.payload.dirty, e.payload.path)
  );

File: src-tauri/src/events.rs
use serde::Serialize;
use tauri::{AppHandle, Emitter};
use crate::domain::model::{Match, Phase, Round, Tournament};
use crate::domain::standings::TeamStanding;

pub const TOURNAMENT_LOADED: &str = "tournament-loaded";
pub const TOURNAMENT_CLOSED: &str = "tournament-closed";
pub const TOURNAMENT_UPDATED: &str = "tournament-updated";
pub const MATCH_UPDATED: &str = "match-updated";
pub const ROUND_GENERATED: &str = "round-generated";
pub const PHASE_CHANGED: &str = "phase-changed";
pub const STANDINGS_UPDATED: &str = "standings-updated";
pub const DIRTY_CHANGED: &str = "dirty-changed";

#[derive(Serialize, Clone)]
pub struct PhaseChangedPayload {
    pub from: Phase,
    pub to: Phase,
}

#[derive(Serialize, Clone)]
pub struct DirtyChangedPayload {
    pub dirty: bool,
    pub path: Option<String>,
}

pub fn emit_tournament_updated(app: &AppHandle, t: &Tournament) {
    let _ = app.emit(TOURNAMENT_UPDATED, t);
    let _ = app.emit(TOURNAMENT_LOADED, t);
}

pub fn emit_tournament_closed(app: &AppHandle) {
    let _ = app.emit(TOURNAMENT_CLOSED, ());
}

pub fn emit_match_updated(app: &AppHandle, m: &Match) {
    let _ = app.emit(MATCH_UPDATED, m);
}

pub fn emit_round_generated(app: &AppHandle, r: &Round) {
    let _ = app.emit(ROUND_GENERATED, r);
}

pub fn emit_phase_changed(app: &AppHandle, from: Phase, to: Phase) {
    let _ = app.emit(PHASE_CHANGED, PhaseChangedPayload { from, to });
}

pub fn emit_standings_updated(app: &AppHandle, standings: &[TeamStanding]) {
    let _ = app.emit(STANDINGS_UPDATED, standings);
}

pub fn emit_dirty_changed(app: &AppHandle, dirty: bool, path: Option<String>) {
    let _ = app.emit(DIRTY_CHANGED, DirtyChangedPayload { dirty, path });
}
